compiler/hash-typecheck/src/ops/typing.rs

```diff
@@ -320,16 +337,22 @@ impl<'gs, 'ls, 'cd, 's> Typer<'gs, 'ls, 'cd, 's> {
             Pattern::Tuple(tuple_pattern) => {
                 // For each parameter, get its type, and then create a tuple
                 // type:
-                let params_id = self.params_of_pattern_params(tuple_pattern, ParamOrigin::Tuple)?;
-                let builder = self.builder();
-                Ok(builder.create_rt_term(builder.create_tuple_ty_term(params_id)))
-            }
-            Pattern::Constructor(constructor_pattern) => {
-                let args_id =
-                    self.args_of_pattern_params(constructor_pattern.params, ParamOrigin::Unknown)?;
+                let params_id = self.args_of_pattern_params(tuple_pattern)?;
                 let builder = self.builder();
-                Ok(builder.create_fn_call_term(constructor_pattern.subject, args_id))
+                Ok(builder.create_tuple_lit_term(params_id))
             }
+            Pattern::Constructor(constructor_pattern) => match constructor_pattern.params {
+                Some(params) => {
+                    // We have params to apply
+                    let args_id = self.args_of_pattern_params(params)?;
+                    let builder = self.builder();
+                    Ok(builder.create_fn_call_term(constructor_pattern.subject, args_id))
```

Comment:
I'm guessing it doesn't actually matter if it is struct/enum? But the naming is quite confusing here

Reply:
Yes it doesn't, and I think it's better to tackle removal of enums in favour of structs/unions first before doing the rest of the constructor pattern stuff..